Added Voltage Monitoring with ADC

aditn opened the pull request:
When cutting key, stepper motor sounds different. Stepper motor was moved to Timer4 from Timer1 and ADC is automatically triggered by Timer1.

Commit:
testing homing bug

Commit:
Fixed homing bug. Flag parameter was made into a defined value in limits.h

Commit:
Working on adding timer for voltage monitoring

Commit:
Working on adding timer for voltage monitoring

Commit:
have a timer interrupt written to get adc values. TODO: set up Auto Trigered ADC conversion.

Commit:
have a timer interrupt written to get adc values. TODO: set up Auto Trigered ADC conversion.

Commit:
Seems to overwrite entire voltage array with the final value checked. Also everything is on Timer0 and seems to slow stepper motor down. ADC AutoTriggering only uses Timer0 and Timer1, both of which GRBL uses. Next move is to try and move the current Timer1 to Timer3 in order to use Timer1 for the ADC AutoTriggering.

Commit:
Seems to overwrite entire voltage array with the final value checked. Also everything is on Timer0 and seems to slow stepper motor down. ADC AutoTriggering only uses Timer0 and Timer1, both of which GRBL uses. Next move is to try and move the current Timer1 to Timer3 in order to use Timer1 for the ADC AutoTriggering.

Commit:
reports correct voltages. Will refactor code

Commit:
reports correct voltages. Will refactor code

ap00f committed:
Refactored Code

ap00f committed:
Refactored Code

Commit:
fixed print statements to work with kiosk motion repo.

Commit:
fixed print statements to work with kiosk motion repo.

Commit:
Added Low Pass Filter for Voltage Monitoring

Commit:
Added Low Pass Filter for Voltage Monitoring

Commit:
removed space

Commit:
removed space

penguinland commented on Makefile:
```diff
@@ -42,10 +42,11 @@ FUSES      = -U hfuse:w:0xd8:m -U lfuse:w:0xff:m
 
 # symbolic targets:
 all:	grbl.hex
+#all:   main.elf
```
Is this supposed to be in here? If not (if it's just commented-out code), remove it.

penguinland commented on Makefile:
```diff
@@ -42,10 +42,11 @@ FUSES      = -U hfuse:w:0xd8:m -U lfuse:w:0xff:m
 
 AVRDUDE = avrdude  $(PROGRAMMER) -p $(DEVICE) -B 10 -F -D
-COMPILE = avr-gcc -Wall -Os -DF_CPU=$(CLOCK) -mmcu=$(DEVICE) -I. -ffunction-sections
+COMPILE = avr-gcc  -Wall -Os -g -DF_CPU=$(CLOCK) -mmcu=$(DEVICE) -I. -ffunction-sections
```
Remove the extra space from before `-Wall`. 

I assume the `-g` adds in debugging strings for GDB (or the arduino equivalent). Is that supposed to be in here? Will the compiled binaries still fit on our chip?

aditn replied:
I don't really believe this should be an issue, but the motor was making a different sound Friday afternoon. This issue is more likely due to missed deadlines by the microprocessor as opposed to using too much space, but I will try without the -g next week.

mdicicco replied:
You could create a new compile line here, maybe called DEBUG, that adds the -g flag

penguinland commented on report.c:
```diff
@@ -378,68 +378,16 @@ void report_counters()
-  // Set read bit to be the special feedback bit (ADC14)
-  // This is bit select 6, plus
-  ADMUX = ((1<<REFS0) + (FVOLT_ADC));
-  ADCSRB = (1<<MUX5_BIT);
```
I admit that I'm not very familiar with GRBL, but I had assumed this code was important. Why remove it all?

aditn replied:
This code originally enabled the ADC and found the voltages at certain pins upon request. Now we keep the ADC enabled at all times and essentially poll for voltage values. When voltage is requested, we output the the most recent findings.

penguinland commented on report.h:
```diff
@@ -63,6 +63,10 @@
 #define MESSAGE_DISABLED 5
 
+// Variable for holding voltage value after ADC
+extern uint32_t voltage_result[5];
```
You've introduced a bunch of magic numbers in here. Instead, make them named constants.

``` c
#define VOLTAGE_CHANNEL_COUNT 5
extern uint32_t voltage_result[VOLTAGE_CHANNEL_COUNT];
```

and in report.c: 

``` c
for (i = 0; i<VOLTAGE_CHANNEL_COUNT; i++){
    printInteger(voltage_result[i]);
    if (i<VOLTAGE_CHANNEL_COUNT - 1)
    printPgmString(PSTR(","));
}
```

edit: and similar changes in system.c

penguinland commented on system.c:
```diff
@@ -57,6 +75,57 @@ ISR(TIMER2_COMPA_vect)
+  // that we have exited this interrupt.
+  TIFR1 |= (1<<OCF1A)|(1<<OCF1B);
+  
+  float beta_LPF = 0.25; // coefficient for low pass filter
```
This should probably be defined elsewhere in a list of constants, but I don't know where (and I might be wrong; maybe here is the right place after all). What do you think?

aditn replied:
I made changes regarding this. Let me know what you think.

penguinland commented on system.c:
```diff
@@ -57,6 +75,57 @@ ISR(TIMER2_COMPA_vect)
+  
+  float beta_LPF = 0.25; // coefficient for low pass filter
+  // Final conversion is a 10 bit value stored in ADC
+  voltage_result[voltage_result_index] = lpf(beta_LPF, (uint32_t)ADC,
```
I don't know what type `ADC` is, but here you cast it to an int, then in `lpf` you cast it to a float, work with the float, and cast back to an int before storing it in `voltage_result` here. What is its most intuitive type? Why isn't it that type all the time?

aditn replied:
The Atmega2560 is an 8-bit microcontroller. The final ADC value is reported over two 8-bit registers (`ADCL`and `ADCH`). The value described by "`ADC`" is a 16 bit value that consists of both `ADCL`and `ADCH`. I first cast `ADC`to a `uint32_t`to extend the 0's. I cast these values to floats in `lpf` before any arithmetic to avoid overflow issues (since they are unsigned values). In `request_voltage`, I use the `printInteger`function to report values. Thus I need to cast back to a `uint32_t`once completing the arithmetic in `lpf`.

penguinland replied:
Make that argument to `lpf` a `uint16_t` (or whatever `ADC` normally is), and you won't need to extend the 0's. Make `lpf` return a float (so `voltage_result` is an array of floats), and cast to an int solely for printing. It'll be slightly faster to execute, and easier to read because you don't need to keep casting types so often. It might also be slightly more accurate, or perhaps slightly less accurate, depending on the typical range of values `ADC` can have. If you made it a double instead of a float, it would definitely be more accurate, though I don't know if the Atmega2560 supports doubles.

ap00f committed:
Made refactoring changes based on Pull request comments

ap00f committed:
Made refactoring changes based on Pull request comments

ap00f committed:
Made refactoring changes based on Pull request comments

ap00f committed:
Made refactoring changes based on Pull request comments

ap00f committed:
fix

Commit:
Added separate line for debug flag in Makefile.
Reduced the number of casts in the ADC_ISR, which should decrease the time spent in the ISR. Now casts to uint32_t when we decide to print.

Commit:
Added separate line for debug flag in Makefile.
Reduced the number of casts in the ADC_ISR, which should decrease the time spent in the ISR. Now casts to uint32_t when we decide to print.

kborer commented on stepper.c:
```diff
@@ -672,14 +672,16 @@ void stepper_init()
+  TCCR4A &= ~((1<<WGM41) | (1<<WGM40));
+  TCCR4A &= ~((1<<COM4A1) | (1<<COM4A0) | (1<<COM4B1) | (1<<COM4B0)); // Disconnect OC4 output
+  // TCCR4B = (TCCR4B & ~((1<<CS42) | (1<<CS41))) | (1<<CS40); // Set in st_go_idle().
+  // TIMSK4 &= ~(1<<OCIE4A);  // Set in st_go_idle().
```
why not delete this commented code instead of changing it?

aditn replied:
It was there originally so I didn't delete it at the time. I can now though.

kborer commented on system.c:
```diff
@@ -57,6 +75,56 @@ ISR(TIMER2_COMPA_vect)
+
+/* This performs a low pass filter on ADC values to reduce impact of noise on a final
+   value */
+float lpf(float beta, uint16_t adc, float final_prev){
```
why not just call this function `low_pass_filter`?

kborer commented on system.h:
```diff
@@ -161,5 +161,9 @@ enum {
 
-
+// Voltage Monitoring
+#define VOLT_NUM 5 // number of sensors we are probing
```
if this constant holds the number of sensors, why not call it `VOLTAGE_SENSOR_COUNT`?

Commit:
made changes to variable/function names

Commit:
made changes to variable/function names

penguinland commented on Makefile:
```diff
@@ -42,10 +42,13 @@ FUSES      = -U hfuse:w:0xd8:m -U lfuse:w:0xff:m
 AVRDUDE = avrdude  $(PROGRAMMER) -p $(DEVICE) -B 10 -F -D
-COMPILE = avr-gcc -Wall -Os -DF_CPU=$(CLOCK) -mmcu=$(DEVICE) -I. -ffunction-sections
+CFLAGS = -Wall -Os -DF_CPU=$(CLOCK) -mmcu=$(DEVICE) -I
+COMPILE = avr-gcc $(CFLAGS) -ffunction-sections
```
I am not good at makefiles, but it kinda looks like `COMPILE` gets defined before we add the `-g` to `CLFAGS` down on line 51 when making the debug target. Have you compiled the debug target and gotten the debugging information included in the hexfile?

aditn replied:
I need to understand more about Makefiles before giving a more detailed response. The short response is that the `debug` mode is applied to `CFLAGS` (the `-g`) before the compile lines executes.

penguinland commented on system.c:
```diff
@@ -57,6 +75,56 @@ ISR(TIMER2_COMPA_vect)
+/* This performs a low pass filter on ADC values to reduce impact of noise on a final
+   value */
+float low_pass_filter(float beta, uint16_t adc, float final_prev){
+  float result = (float)final_prev - (beta * (final_prev - (float)adc));
```
No need to cast `final_prev` to a float on this line.

penguinland commented:
On the whole, this LGTM!

mdicicco commented:
@aditn hold off on merging.  Alan's proposed changes are all good, but I still want to take a look before we move forward.

Commit:
Removed redundant float cast

Commit:
Removed redundant float cast

Commit:
Added ADC interrupt disable fns

Commit:
need to test timer changes

Commit:
Merge branch 'moveTimers' of https://github.com/aditn/grbl into moveTimers

Conflicts:
	system.c
	system.h

penguinland commented:
I don't know when this should be merged in, but there are conflicts to resolve before that happens.

mdicicco commented:
no huge rush for this merge

Commit:
Removed low pass filtering to improve timing

Commit:
minor changes to remove extra print statements

aditn commented:
This pull request is being closed because another pull request will include code from here.
